### ISSUE TITLE
Project 4

### ISSUE THREAD

**Issue by drkostas** [drkostas/COSC522#2]
**The objective of this project is to have a thorough understanding of the traditional multi-layer perceptron and backpropagation.**

Dataset: **MNIST**

Tasks:

- [X] Create MultiLayerPerceptron Class
- [X] Make it work for the XOR data
- [X] Make it work for the MNIST dataset
- [x] Add functionalities
    - [X] sigmoid
    - [x] ReLU
    - [x] softmax
    - [x] tanh
    - [X] Gradient descent
    - [X] Stochastic and batch Gradient Descent
    - [X] Shuffle
    - [X] Normalize dataset
    - [X] Standarize dataset
    - [x] L1 standarization
    - [x] L2 standarization
    - [x] momentum
    - [X] accuracy
    - [X] Cross Entropy/Log Loss
    - [x] Time per epoch and total
    - [x] Early Stopping
- [X] Gather results for different
    - [X] number of layers
    - [X] learning rates
    - [X] Batch Sizes
- [X] Visualize:
    - [X] Accuracy
    - [X] Losses
    - [X] Training Time (total and per epoch)
- [X] Clean code/remove debugging prints/organize notebook
- [X] Create report and mention: that you created code from scratch, used set seed, added extra functionalities
